### ISSUE TITLE
Textpattern 4.9.0 release flight plan

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1903]
We are targeting Textpattern 4.9.0 for release in Q2 2024 to introduce new features & improvements, provide initial support for PHP 8.3, address some snags, and provide further refinements. The release flight plan includes at least one beta release before the production release.

* [4.9.0 branch](https://github.com/textpattern/textpattern/tree/v4.9.0)

Please update this post and subsequent replies with tasks etc as needed, assigning yourself or others where appropriate.

**Timeline**

* beta release(s): Q1 2024 or Q2 2024
* GA release: Q2 2024

# Beta release(s)

## Personnel

* Release pilot: tbc
* Release co-pilot: tbc

## Preflight

- [ ] preflight
  - [ ] latest translations (@philwareham)
  - [ ] latest pophelp (@philwareham)
  - [ ] draft .com blog post (release pilot, or assigned person)
  - [ ] write `HISTORY.txt` (everyone)
    - [ ] additional credits where not already listed (e.g. patched security issues) (everyone)
  - [ ] top and tail `HISTORY.txt` with additional information (release pilot or co-pilot)
  - [ ] tag public/admin themes with release tag (@philwareham)
  - [ ] tag Textpacks and Pophelps with release tag (@philwareham)

## Take Off

- [ ] take off
  - [ ] package up for GitHub and .com (release pilot)
  - [ ] update demo site (@petecooper)
  - [ ] publish .com blog post (release pilot, or assigned person)
  - [ ] forum announcement (release pilot, or assigned person)
  - [ ] forum feedback (release pilot, or assigned person)

## Inflight service

- [ ] inflight service
  - [ ] socials (@petecooper or @philwareham)

**Repo Issues**

# GA release

## Personnel

* Release pilot: tbc
* Release co-pilot: tbc

## Preflight

- [ ] preflight
  - [ ] latest translations (@philwareham)
  - [ ] latest pophelp (@philwareham)
  - [ ] draft .com blog post (release pilot, or assigned person)
  - [ ] write `HISTORY.txt` (everyone)
    - [ ] additional credits where not already listed (e.g. patched security issues) (everyone)
  - [ ] top and tail `HISTORY.txt` with additional information (release pilot or co-pilot)
  - [ ] tag public/admin themes with release tag (@philwareham)
  - [ ] tag Textpacks and Pophelps with release tag (@philwareham)

## Take Off

- [ ] take off
  - [ ] package up for GitHub and .com (release pilot)
  - [ ] update `latest.zip` and `latest.tar.gz` in server config (@petecooper)
  - [ ] update language pack links in server config (@petecooper)
  - [ ] update demo site (@petecooper)
  - [ ] publish .com blog post (release pilot, or assigned person)
  - [ ] forum announcement (release pilot, or assigned person)
  - [ ] forum feedback (release pilot, or assigned person)

## Inflight service

- [ ] inflight service
  - [ ] close milestone (release pilot, or assigned person)
  - [ ] update `TXPversion` on `rpc.textpattern.com` (release pilot, or assigned person)
  - [ ] update https://en.wikipedia.org/wiki/Template:Textpattern_version (@petecooper or @philwareham)
  - [ ] socials (@petecooper or @philwareham)
  - [ ] update https://textpattern.com/about/system-requirements to reflect changes

**Comment by petecooper** [comment]
Can we pin down and agree on the milestone for the first beta of 4.9.0, please?

The 4.9 shopping list milestone is here: https://github.com/textpattern/textpattern/milestone/5

Remember that we don't have to get all this into the 4.9.0 release, enhancements are totally fine in patch releases.

We're more than 2 years since a release and it will do us a lot of good to get 4.9.0 out there, especially given the latest PHP we officially support (8.1) is now in its security support era.

Paging @Bloke @bloatware @philwareham for commentary.

**Comment by Bloke** [comment]
I'd love to get the constraint going but I'm not delaying any longer. If I can crack it then it might squeeze into 4.9 final. Otherwise it'll be 5.0.0.

The only outstanding things from my perspective both relate to the CSS for the plugins panel. See https://github.com/philwareham/textpattern-hive-admin-theme/issues/683 and https://github.com/philwareham/textpattern-hive-admin-theme/issues/669.

With recent changes to the plugin panel Manage column there's also a slight annoyance that if one of the elements is missing and is injected by JS, you get an empty double pipe symbol. Solutions welcomed.

**Comment by philwareham** [comment]
Hi. I'll try to mop up those CSS issues next week for you. I'll also pull in any recent translations.

**Comment by bloatware** [comment]
There are some other css to tweak, namely for the anchors that have been converted to buttons.

**Comment by Bloke** [comment]
> Hi. I'll try to mop up those CSS issues next week for you. I'll also pull in any recent translations.

Awesome, thank you. Btw, I'm not suggesting my PRs are the correct or best way to do this, so if anyone has any other ways to achieve this - maybe even refactor the way the Manage column is habdled - then by all means go for it.

The main issue with [669](https://github.com/philwareham/textpattern-hive-admin-theme/issues/669) that I'm trying to address can be seen by installing a plugin that responds to lifecycle events and has prefs and docs.

Install it and take a look on the Plugins panel while it's not activated and you'll see:

```
Help |             | Export
```
(and maybe `Upgrade to vA.B.C` if it's available on the plugins site)

When you enable the plugin, `Options` pops up in the central empty slot. It'd be nice if it didn't leave a gap. But if we just go for the PR as is, then it will render two pipes next to one another, which is equally rubbish.

I'm not sure how to fix it, beyond getting rid of the hard coded pipes and maybe rendering the things as buttons, or msrking up thr oipes somehow so they collapse automatically if two occur in a row.

**Comment by philwareham** [comment]
I think I've done all the CSS requests you wanted now. Fixed various styling issues with links that had recently been turned into buttons, added Stef's bits that fix the plugin panel options links. Let me know if you spot anything else that needs my attention.

**Comment by philwareham** [comment]
@bloatware just check my dev install of Textpattern and I hadn't even noticed that plugin uploader widget - I guess it has been added recently?

Will try to tame it soon.

**Comment by philwareham** [comment]
Just need some direction on the plugin page changes, so I have run the latest dev build and this is what is there:
<img width="1528" alt="Screenshot 2024-03-01 at 16 57 05" src="https://github.com/textpattern/textpattern/assets/413665/95eb9639-6a76-4cd1-9974-86d9c8736dd4">

 Whereas in my (older) mockups it was like this:
<img width="1528" alt="Screenshot 2024-03-01 at 16 58 19" src="https://github.com/textpattern/textpattern/assets/413665/b9a4fe81-9855-4583-a539-ce4271ef3261">

A few questions:

1. Is the 'Import from disk' option actually available any more? If so, what process makes that appear as I can't remember the use case?
2. We'll need some division between the file upload UI and the textbox install UI (and maybe different wording on the action buttons to separate their contexts).
3. Not too keen on the twisty header to show the UI, I would rather show all but fix the layout issues.
4. I need to incorporate the edit button into the second install section (the textarea one) yes? I assume the edit is for this rather than on the file uploader (or can not be used on either?).

Thanks for any feedback. Just need to know the UX before I do the UI.

**Comment by phiw13** [comment]
That “Edit” button should really be labelled “New Plugin”, because that is what it does, open a blank (new) plugin editor page.

And yes the “Import from disk” option still exists when a plugin exist on the file system that is not registered/installed.

So the are actually four possible actions: Create new, Upload(zip or php format), Import and Install (txt file, the oldest  way to install plugin files.

Agree on the twisty header stuff, after actually using the plugin panel.

**Comment by petecooper** [comment]
I'm keen to get 4.9.0 out of the door so we can move to 5.0.

What's in the work queue for a 4.9.0 beta, and what do you need from me?

Tagging @bloatware @Bloke @philwareham

**Comment by bloatware** [comment]
I've got no other plans than some optimizations and bugfixes. Test test test..

**Comment by Bloke** [comment]
@philwareham Are you done with the plugin panel changes in the top area? I haven't checked the latest dev in a few weeks.

**Comment by philwareham** [comment]
No, sorry, I haven't had a chance to look at it yet. Promise I will do that early next week as I realise this is blocking your release schedule.

**Comment by Bloke** [comment]
Awesome, thank you.

**Comment by Bloke** [comment]
Sorry I've not been about much. Stupidly under the weather and it's kicking me big time.

Can we release the beta yet?

**Comment by petecooper** [comment]
Paging @bloatware @philwareham. @Bloke has requested I push the buttons for beta 1 - please let me know if you're happy for me to proceed. I will wait to hear back from both of you before I take any action. Thank you.

**Comment by philwareham** [comment]
Fine by me - let's go for it. I can do the translations sync when you are ready to release.

**Comment by bloatware** [comment]
I'm available for about a week if needed, but then will mostly sail away until the end of the summer.

Leave me this evening please to recall whether something urgent remains on todo.

**Comment by petecooper** [comment]
I've started an article for our intentions with 4.9 (article ID 484), I'll have some time to finish it later this week.

I'll pencil in beta 1 for w/c July 29th 2024 after @bloatware feedback.

**Comment by Bloke** [comment]
Thanks Pete. Even if we have missed anything we can mop it up before final release.

**Comment by petecooper** [comment]
https://textpattern.com/weblog/textpattern-cms-490-will-support-php-84-and-mysql-84

**Comment by petecooper** [comment]
Barring _force majeure_, first beta of 4.9.0 will hit the streets today (Aug 12 2024).

Paging @Bloke @philwareham (and @bloatware also, though I know he's sailing away).

**Comment by petecooper** [comment]
First beta is live: https://github.com/textpattern/textpattern/releases/tag/4.9.0-beta

Some snags with the toolbelt instructions, but I'll hop into issues to flag the ones I can't fix myself.

I will announce _unofficially_ on the forum for the keen folks to test, and announce _officially_ by Thursday if stuff hasn't exploded with early bird testers,

**Comment by petecooper** [comment]
https://beta-demo.textpattern.co

OK, something's not right…that doesn't match `dev` and there're language pack issues. I think something's come unstuck with the toolbelt instructions and the changes over the last couple of years.

@Bloke - would you have any capacity for beta.2 technicals please, if I've screwed up the release for this beta? I can do the words if that helps.

**Comment by Bloke** [comment]
No worries. There have been some changes so a few things in the toolbelt may be wonky. I'll see if I can track anything down.

**Comment by petecooper** [comment]
Thanks @Bloke - much appreciated

**Comment by Bloke** [comment]
Beta.1 launched.

I cheated a bit and reused the same file IDs on textpattern.com and just updated the file checksums accordingly. Same with the beta release announcement on GitHub - I reused Pete's one and just changed 'beta' to beta.1'.

Also fixed up a few things in the toolbelt, but we do need to address the txp-gitdist.sh shell script because it doesn't verify the sha checksums (see https://github.com/textpattern/textpattern-toolbelt/issues/30).

We have a rogue `release-4.9.0-beta` branch on the git repo so we'll need to delete that and resync our local branches accordingly. I'm never sure if it's better to delete the branch locally and `push origin --delete`, or delete it from GitHub and pull down to our local dev environments. Maybe it doesn't matter. See [Step 19](https://github.com/textpattern/textpattern-toolbelt?tab=readme-ov-file#step-19-tidy-up-branches) for details. Lots could go wrong with a misplaced command and I'm not brave enough to do it with eyes this tired!

The [Getting Started](https://textpattern.com/start) doc has been updated to expose (unhide) the Beta section, but there's (as yet) no .com article and I've run out of flying hours, sorry.

**Comment by phiw13** [comment]
Maybe a problem?

Trying to “update” a site that is running 4.9.0-dev latest to 4.9beta.1 (by replacing all relevant files from the git pull to those in the beta package)

```
Parse error: syntax error, unexpected token "<<", expecting end of file in /Users/[username]/Sites/txpdev/textpattern/index.php on line 30
```

I don’t remember this happening in the past.

**Comment by Bloke** [comment]
Oh balls, I missed a merge conflict in /textpattern/index.php. it'll need rebuilding. Hang fire please.

**Comment by Bloke** [comment]
Alright, re-re-issued. Note to self: don't merge late at night.

Thanks for your patience and sorry for the false starts.

**Comment by phiw13** [comment]
4.9.-dev -> 4.9 beta: **success** ( 4.8.8 -> 4.9 beta will have to wait).

>  Note to self: don't merge late at night.

Nah… Blame the Olympics or some such!

**Comment by phiw13** [comment]
Ok:

* 4.9-dev -> 4.9-beta.1a: all works fine both admin side and public side
* 4.8.8 -> 4.9-beta.1x: appears to work on both admin and public side with one problem: on _every_ page load of the public side I get the “please force reload or clear your browser caches” message. There might be something on my side, though. And BTW, that is a very minimal site  on local host, basically the default install.

**Comment by Bloke** [comment]
> please force reload or clear your browser caches

I've seen that occasionally and never got to the bottom of it. Seems some sort of aggressive caching combined with 'version' pref not being set properly in dev environments triggers it, maybe?

I'll triple check in case there's a remaining issue with the bundle version numbers, but I"m pretty sure I caught them all this time. textpattern.js is usually the culprit.

**Comment by petecooper** [comment]
- I'll have an additional pair of eyes on things later today, too. I'll draft a beta release article for .com, too.

**Comment by petecooper** [comment]
Done: https://textpattern.com/weblog/textpattern-490-beta1-released-php-84-mysql-84-features-fixes-and-more

**Comment by petecooper** [comment]
Now we're in a [PHP 8.4](https://www.php.net/releases/8.4/en.php) world, I'll take a look at any low hanging fruit for 4.9.0.

Current Textpattern 4.9.x release issues: https://github.com/textpattern/textpattern/issues?q=is%3Aopen+is%3Aissue+milestone%3Av4.9

I've made a Textpattern 4.9.0-specific milestone so we can track the items needed for launch: https://github.com/textpattern/textpattern/issues?q=is%3Aopen+is%3Aissue+milestone%3Av4.9.0

Can we sift through the 4.9.x issues and see what's viable for 4.9.0, and what can wait?

**Comment by bloatware** [comment]
I see nothing mandatory in the remaining issues. Hive themes still need some tweaks, though, since the anchors in `#tab-preview` and `#tab-html` are buttons now.

**Comment by petecooper** [comment]
Requesting a thumbs up or thumbs down on the feasibility / usefulness of a second beta release.

@bloatware @Bloke @philwareham

**Comment by Bloke** [comment]
I'm happy to go straight to RC. There's quite a few nice bits and pieces in it since beta-1 and I was hoping to get time to put some stuff in to help themers bootstrap their code, but I haven't. Might have to wait for txp 5. If we go to beta-2 then I may be able to squeeze some of that in but no guarantees.

**Comment by bloatware** [comment]
> I see nothing mandatory in the remaining issues. Hive themes still need some tweaks, though, since the anchors in `#tab-preview` and `#tab-html` are buttons now.

Statu quo. Beta-2 wouldn't hurt, perhaps.

**Comment by Bloke** [comment]
Fine by me.

**Comment by petecooper** [comment]
I'll have capacity to run the beta.2 release late December, if that fits.

**Comment by petecooper** [comment]
Requesting a thumbs up or thumbs down on viability of beta.2 (I am happy to run it or assist if someone else wants to take the reins) in the next 7 days.

Tagging @Bloke @bloatware @phiw13 @jools-r for commentary. Anyone else I've forgotten?

**Comment by bloatware** [comment]
@philwareham ? Hive themes still need some tweaks iirc.

**Comment by philwareham** [comment]
Hi, I will have 2-3 hours I can spare before I get fully back in the swing of work for 2025 next week. Can you please just quickly summarise what I need to look at in the themes to get them ready, just so I can use the time as best I can? I will also look at the current translations status too and get any new strings into the workflow. Again, if any new strings can be listed either here or in the Textpack/Pophelp repos that will help me greatly. Cheers and happy new year all.

**Comment by bloatware** [comment]
Thanks @philwareham and happy new year. Essentially, only preview widgets need to be updated:

- `#view_modes a` rules (in body/excerpt previews) need to be transformed into `#view_modes button`. I've tried straightforward replacements, but the result was not top.
- The `Sandbox` label in the full article preview needs some margins.
- Probably, `.ui-dialog .ui-dialog-content` containing an `iframe` does not need (so much) padding.

@phiw13 certainly knows how to fix it quickly.

**Comment by phiw13** [comment]
@philwareham 

I am traveling for the holidays, but as far as the few features mentioned by Oleg, you can go checking what I did for Sandspace
Mockup of write panel (the iframe based preview - button at the top - does not work in the mockup, I can‘t be bothered…):
https://dev.l-c-n.com/txp-4.9/txp-content-write.html

Download theme: 
https://l-c-n.com/stuff/sandspace-a-theme-for-the-textpattern-cms-admin-interface#txp-download

For the iframe based preview inside the `ui-dialog`, search for `iframe`in the stylesheet:
https://dev.l-c-n.com/txp-4.9/phw_sandspace/sndsp.css

-^-

Other issue to consider: integrate / use the core custom properties in your theme (see the core [textpattern.css)](https://github.com/textpattern/textpattern/blob/dev/textpattern/textpattern.css)

Lazy way: alias them  – give the hive property the value of the core property, e.g for the main text color: `--clr-text: var(--txp-primary-text)`

Longer way: replace usage everywhere in the theme – the property names used in the core stylesheet are self-evident, the values are taken from theme hive

**Comment by petecooper** [comment]
@philwareham If you have any capacity to sanity check https://github.com/textpattern/textpattern/issues/1956 (DOMPurify not landing correctly with npm) that would be greatly appreciated. Thank you!

**Comment by petecooper** [comment]
Proposal: we release `beta.2` in the coming weeks with latest changes and then address the remaining snags & to-do items, it'd be useful to get a fresh cut from that branch.

Tagging @bloatware @Bloke @philwareham @phiw13 @jools-r for any commentary plus a 'Go' or 'No Go'. Thank you.

**Comment by Bloke** [comment]
I'm all for that. I've got time next week to bundle a release.

**Comment by petecooper** [comment]
Can we safely defer https://github.com/textpattern/pophelp/milestone/4 until after beta.2?

**Comment by Bloke** [comment]
> Can we safely defer https://github.com/textpattern/pophelp/milestone/4 until after beta.2?

Yeah, easy to add if I get a moment while @bloatware has a think about [a possible tweak](https://forum.textpattern.com/viewtopic.php?id=52532) to improve (or ditch) the "someone modified this article while you weren't watching" notification. If we don't get to the strings before that's done, they can be folded in shortly after beta.2.

And thank you for the article text. Awesome stuff.

**Comment by petecooper** [comment]
I've made a PR for each theme to bump the year to 2025.

@philwareham Have you anything to include in the beta.2 release on the themes front?

**Comment by Bloke** [comment]
> I've made a PR for each theme to bump the year to 2025.

Yeah I did that by hand for the release I started. And also [omitted the config-dist](https://github.com/textpattern/textpattern/commit/c5e107e80e16d2a9dc7ce65388064ba18d33d1d9) as I see you've now done (thank you).

**Comment by petecooper** [comment]
>Are we all good for me to pull the trigger on beta.2 today?

Y from me. I'm around for co-pilot duties.

**Comment by petecooper** [comment]
Polite request: please use https://github.com/textpattern/textpattern-toolbelt/blob/main/release/txp-gitdist-xz.sh for the release archives as that creates the .tar.xz flavour and we can close https://github.com/textpattern/textpattern/issues/1942 if that works as intended. Thank you.

**Comment by petecooper** [comment]
PrismJS kerfuffle appears resolved, shall we shoot for a release this week, @Bloke? I can make myself available if & when you're ready.

**Comment by Bloke** [comment]
Woot. [Beta.2 finally out](https://github.com/textpattern/textpattern/releases/tag/4.9.0-beta.2). Thank you ever so much for all your hard work on this, and apologies for the delays building time in to get it out the door. Real life, ugh.

I don't envisage much fallout from this one so hopefully we can move to final within the next handful of weeks. Then get moving properly on Txp 5!

**Comment by petecooper** [comment]
beta.2 OP items archived away, onwards to RC or GA!

Thank you @Bloke for pushing the buttons.

**Comment by jools-r** [comment]
Well done! Thanks to everyone. We're getting there.

**Comment by petecooper** [comment]
Warming this up…

Do we need a beta.3 or RC release prior to GA? We skipped RC for 4.8.0, though it did have 3x beta releases.

Tagging @Bloke @bloatware @philwareham @jools-r for advice.

**Comment by etc** [comment]
> [@etc](https://github.com/etc) The latest changes to the article workflow have resolved the issue with spurious notifications. Thank you so much. And thanks for rolling in the latest translations [@philwareham](https://github.com/philwareham)
> 
> Are we all good for me to pull the trigger on beta.2 today?

I am not involved in this project.

**Comment by bloatware** [comment]
It [looks](https://github.com/textpattern/textpattern/issues/1986?notification_referrer_id=NT_kwDOALCsQ7QxNTc4NDY5MjQ3NDoxMTU3ODQzNQ) like few strings are still missing? Once added, I think we can go RC.

**Comment by petecooper** [comment]
Added RC 1 to OP.

**Comment by petecooper** [comment]
First draft of rc.1 post done, ID 487.

**Comment by Bloke** [comment]
Are we bothering with RC? Do we historically get more eyeballs on them compared to beta?

As far as I can see we have:
* A few strings to mop up (https://github.com/textpattern/textpattern/issues/1986 among others maybe?)
* The `br` trailing quote thing to decide upon (easy).
* Further head-scratching on the [broken headers in some installs](https://forum.textpattern.com/viewtopic.php?pid=339319#p339319) and see commit 9a20ce84.
* https://github.com/textpattern/textpattern/issues/1984
* Possible quick thoughts on allowing core to read pophelp files from a plugin's directory (see https://forum.textpattern.com/viewtopic.php?id=52559). Suggest the same naming convention we have in our core /textpattern/lang directory.

Happy toskip it and go straight to release but if y'all think an RC has value, then cool.

**Comment by petecooper** [comment]
I could have sworn the consensus was RC. Now I'm having doubts. I'll check.

**Comment by petecooper** [comment]
Reheating this once more for RC or GA.

We have a couple of issues tagged with 4.9.0: https://github.com/textpattern/textpattern/milestone/26

- Can we push these to 4.9 (i.e., not 4.9.0) milestone and proceed to RC or GA?

@bloatware @Bloke Are there any aspects you're working on that are close to ready and you want in the next release?

From my side, I think GA might be sensible, and we can address the outstanding issues in a patch release or three – and then agree on 5.0 roadmap.

Tagging @philwareham for design & other feedback, and @jools-r for overall insights.

Thank you.

**Comment by Bloke** [comment]
@jools-r It does smell like a bug so if you can throw a PR together that would be amazing, thank you.

**Comment by petecooper** [comment]
- Thanks @jools-r & @Bloke, I've assigned you both to #1991.

**Comment by bloatware** [comment]
I would like to fix #1987 (and #1992 if time permits) before GA. It would be nice to avoid (as much as possible) patching 4.9 in parallel with 5.0.

**Comment by petecooper** [comment]
Thank you @bloatware – I've tagged https://github.com/textpattern/textpattern/issues/1987 and https://github.com/textpattern/textpattern/issues/1992 with `4.9.0` milestone.

**Comment by petecooper** [comment]
Howdy, sirs. Checking in - what's the latest with 4.9.0 tasks, please?

I'll take point on some of the non-dev admin stuff (release article, etc), but I think we'd all benefit from a GA release some time soon-ish, then we can sort the 5.0 roadmap.

Thank you, and best wishes to you from sunny Cornwall, UK!

**Comment by jools-r** [comment]
🎉🎉🎉

**Comment by bloatware** [comment]
Thanks @petecooper, it looks like the remaining issues mainly concern language packs, where I'm of little help. Xmas release would be great!

**Comment by petecooper** [comment]
PHP 8.5.0 has landed and is compiling.

I have had a first pass at changing .com article 487 from an RC to a GA release, I'll take a second pass later today.

**Comment by Bloke** [comment]
I'm being really naughty and trying to squeeze in a fairly big change but it's totally backwards compatible and seems to be holding up well. If it doesn't work or is deemed too risky at this stage I'll delay it to a later release. I want it committed today if I can so people can at least decide if it's worth doing in 4.9.0.

**Comment by jools-r** [comment]
Neat! I gave it a quick whirl with my test image thumbnail generation form and it worked nicely:

```
<txp:variable name="srcset" value="400|4x3, 800|3x3, 1200|2x3" breakby="," break="," wraptag='srcset="<+>"' output trim>
<txp:image_url thumbnail width='<txp:yield item breakby="|" limit="1" />' crop='<txp:yield item breakby="|" offset="1" />' /> <txp:yield item breakby="|" limit="1" />w
</txp:variable>
```

which (in a `txp:images` loop) produces this, which is spot on (except for the extraneous spaces):

```
srcset=" http://newimage.test/images/thumb/w400-c4x3/images/2.jpg?token=XXX 400w , http://newimage.test/images/thumb/w800-c3x3/images/2.jpg?token=YYY 800w , http://newimage.test/images/thumb/w1200-c2x3/images/2.jpg?token=ZZZ 1200w "
```

A question: Am I going to run into problems with yields inside yields? I'm thinking of a shortcode situation where you might just specify the desired widths in, say, a `srcset="400, 800, 1200"` attribute of shortcode, and then use that for `value='<txp:yield name="srcset" />'` in the `txp:variable tag.

**Comment by philwareham** [comment]
All themes have been tagged with 4.9.0 now so I think my part is done. Let me now though. Cheers!

**Comment by Bloke** [comment]
It's becoming solider by the minute. I'm happy with it.

Left to do (today, now):

1. Expose a couple of the more useful defaults as config.php options instead of hard-coding them.
2. Add a callback to imageBuildURL() so plugins can override the URL (e.g. to add SLIR functionality like background colour support or progressive saving, that we don't support out of the box).

EDIT: oh, and @bloatware is going to look at the tag consistency between `<txp:image thumbnail>` and `<txp:article_image>`

**Comment by Bloke** [comment]
I say GO.

If I have time later I might copy across the ARIA work I did on the Articles panel to the others, because it's helpful for assistive tech (and the strings exist now, so we might as well use them). But if I don't make it, that won't be a deal breaker for release because the other panels are ancilliary compared to the Articles panel.
